beluga/include/beluga/sensor/data/occupancy_grid2_mixin.hpp

```diff
@@ -153,31 +130,59 @@ class BaseOccupancyGrid2 : public BaseLinearGrid2<Derived> {
-    return cells | ranges::views::transform(
-                       [this, frame](const auto& cell) { return this->self().coordinates_at(cell, frame); });
+    const auto get_coordinates = [this, frame](const auto& cell) { return this->self().coordinates_at(cell, frame); };
+    return cells | ranges::views::transform(get_coordinates);
   }
 
-  /// Retrieves range of free grid cell indices.
+  /// Retrieves a range containing the ids of free cells.
   [[nodiscard]] auto free_cells() const {
-    return ranges::views::enumerate(this->self().data()) |
-           ranges::views::filter([value_traits = this->self().value_traits()](const auto& tuple) {
-             return value_traits.is_free(std::get<1>(tuple));
-           }) |
-           ranges::views::transform([](const auto& tuple) { return std::get<0>(tuple); });
+    const auto is_free = [value_traits = this->self().value_traits()](const auto& tuple) {
+      return value_traits.is_free(std::get<1>(tuple));
+    };
+    const auto get_cell_id = [](const auto& tuple) { return std::get<0>(tuple); };
+    return this->self().row_major_data() | ranges::views::filter(is_free) | ranges::views::transform(get_cell_id);
   }
 
-  /// Retrieves grid data using true booleans for obstacles.
+  /// Retrieves a range containing the cells that are obstacles only.
   [[nodiscard]] auto obstacle_data() const {
-    return this->self().data() |
-           ranges::views::transform([value_traits = this->self().value_traits()](const auto& value) {
-             return value_traits.is_occupied(value);
-           });
+    const auto is_occupied = [value_traits = this->self().value_traits()](const auto& tuple) {
+      return value_traits.is_occupied(std::get<1>(tuple));
+    };
+    const auto get_cell_id = [](const auto& tuple) { return std::get<0>(tuple); };
+    return this->self().row_major_data() | ranges::views::filter(is_occupied) | ranges::views::transform(get_cell_id);
+  }
+
+  using Mixin::data_at;
+
+  /// Gets cell data, if included.
+  /**
+   * \param xi Grid cell x-axis coordinate.
+   * \param yi Grid cell y-axis coordinate.
+   * \return Cell data if included, `std::nullopt` otherwise.
+   */
+  [[nodiscard]] auto data_at(int xi, int yi) const {
+    return this->self().contains(xi, yi) ? std::make_optional(this->self().cell(xi, yi)) : std::nullopt;
+  }
+
+ private:
+  /// delivers a range of tuples with cell and cell value
+  [[nodiscard]] auto row_major_data() const {
+    const auto width = this->self().width();
+    const auto height = this->self().height();
+    const auto cell_count = width * height;
+    const auto to_raster_data = [this, width](std::size_t raster_index) {
+      const auto xi = static_cast<int>(raster_index % width);
+      const auto yi = static_cast<int>(raster_index / width);
+      const auto cell = Eigen::Vector2i(xi, yi);
+      return std::make_tuple(cell, this->self().data_at(cell).value());
```

Comment:
@glpuga meta: this assumes the underlying grid is at least dense (by rasterizing w/o checking for data availability). FWIW one of the reasons why at the time I refrained from using mixins for the grid hierarchy are these dependencies. They aren't really independent (although I like that we split dense and regular, that was unnecessary).